pkg/services/ngalert/notifier/receiver_svc.go

```diff
@@ -232,3 +323,89 @@ func (rs *ReceiverService) decryptOrRedact(ctx context.Context, decrypt bool, na
+func (rs *ReceiverService) getContactPointProvenance(ctx context.Context, r *definitions.PostableApiReceiver, orgID int64) (models.Provenance, error) {
+	if len(r.GrafanaManagedReceivers) == 0 {
+		return models.ProvenanceNone, nil
+	}
+
+	storedProvenances, err := rs.provisioningStore.GetProvenances(ctx, orgID, (&definitions.EmbeddedContactPoint{}).ResourceType())
+	if err != nil {
+		return "", err
+	}
+
+	// Current provisioning works on the integration level, so we need some way to determine the provenance of the
+	// entire receiver. All integrations in a receiver should have the same provenance, so we use the first one and
+	// ensure all others match.
+	firstProvenance := models.ProvenanceNone
+	if p, exists := storedProvenances[r.GrafanaManagedReceivers[0].UID]; exists {
+		firstProvenance = p
+	}
+	for _, contactPoint := range r.GrafanaManagedReceivers {
+		provenance := models.ProvenanceNone
+		if p, exists := storedProvenances[contactPoint.UID]; exists {
+			provenance = p
+		}
+		if provenance != firstProvenance {
+			return "", fmt.Errorf("all integrations must have the same provenance")
+		}
+	}
+	return firstProvenance, nil
+}
+
+func (rs *ReceiverService) checkOptimisticConcurrency(current *definitions.PostableApiReceiver, provenance models.Provenance, desiredVersion string, action string) error {
+	if desiredVersion == "" {
+		if provenance != models.ProvenanceFile {
+			// if version is not specified and it's not a file provisioning, emit a log message to reflect that optimistic concurrency is disabled for this request
+			rs.log.Debug("ignoring optimistic concurrency check because version was not provided", "receiver", current.Name, "operation", action)
+		}
+		return nil
+	}
+	currentVersion := calculateReceiverFingerprint(current)
+	if currentVersion != desiredVersion {
+		return ErrVersionConflict.Errorf("provided version %s of receiver %s does not match current version %s", desiredVersion, current.Name, currentVersion)
+	}
+	return nil
+}
+
+// getReceiverByUID returns the index and receiver with the given UID.
+func getReceiverByUID(cfg definitions.PostableUserConfig, uid string) (int, *definitions.PostableApiReceiver) {
+	for i, r := range cfg.AlertmanagerConfig.Receivers {
+		if getUID(r) == uid {
+			return i, r
+		}
+	}
+	return 0, nil
+}
+
+// getUID returns the UID of a PostableApiReceiver.
+// Currently, the UID is a hash of the receiver name.
+func getUID(t *definitions.PostableApiReceiver) string { // TODO replace to stable UID when we switch to normal storage
+	sum := fnv.New64()
+	_, _ = sum.Write([]byte(t.Name))
+	return fmt.Sprintf("%016x", sum.Sum64())
+}
+
+// TODO: Check if the contact point is used directly in an alert rule.
+// isReceiverInUse checks if a receiver is used in a route or any of its sub-routes.
+func isReceiverInUse(name string, routes []*definitions.Route) bool {
```

Comment:
What about notification policies in rules?

Reply:
I'll add in a followup. The solution should probably be shared with time interval deletion as well.